### PR TITLE
Show status of LSP actions

### LINKED ISSUE #4380: Find all references progress
### Check for existing issues

- [X] Completed

### Describe the feature

A progress bar or a progress log for when `Find All References` is triggered. Currently there's no progress bar when looking for references. Sometimes I can take minutes for it to find the references. A progress bar or a progress log (like when a project is being loaded and rust-analyzer runs) would be nice.

### If applicable, add mockups / screenshots to help present your vision of the feature

Log line like the following on the left down side with the reference we are searching for + some progress bar.

image

### PR DESCRIPTION
**Pull request by xyzqm**
Fixes #4380

Parts im still unsure about:
- [x] where exactly I should call `on_lsp_start`/`on_lsp_end`
- [x] how to handle things better than `let is_references = TypeId::of::<R>() == TypeId::of::<GetReferences>();`, which feels very janky
- [x] I want to have the message be something like `"Finding references to [...]"` instead of just `textDocument/references`, but I'm not sure how to retrieve the name of the symbol that's being queried
- [ ] I think the bulk of the runtime is occupied by `let result = language_server.request::<R::LspRequest>(lsp_params).await;`, but since `ModelContext` isn't passed into it, I'm not sure how to update progress from within that function
- [x] A good way to disambiguate between multiple calls to the same lsp function; im currently using the function name itself as the unique identifier for that request, which could create issues if multiple `textDocument/references` requests are sent in parallel

Any help with these would be deeply appreciated!

Release Notes:

- Adds a status indicator for LSP actions

### CODE REVIEW

**Comment by TerminalFi** [comment]
@danielzsh LEO request and response contain a unique ID. this is generally how the LSP client can track a response to a request

**Comment by TerminalFi** [comment]
For point 3

you’ll probably have to do a query for what is under the cursor in the active buffer to get what is being queried since the actual LSP request only contains Position information of where the cursor is at time of request

**Comment by danielzsh** [comment]
> @danielzsh LEO request and response contain a unique ID. this is generally how the LSP client can track a response to a request

Do you know where I can get that unique ID? I can't find it anywhere in the codebase 😰

**Comment by TerminalFi** [comment]
> > @danielzsh LEO request and response contain a unique ID. this is generally how the LSP client can track a response to a request
> 
> 
> 
> Do you know where I can get that unique ID? I can't find it anywhere in the codebase 😰 

lsp.rs:122

**Comment by TerminalFi** [comment]
I think that would then create 2 places that ids would have to be managed and might not be the ideal approach. But will let a Zed dev provide feedback. 

Ideally you'll expose it from the language server. Since it is already associated with a buffer.

**Comment by mikayla-maki** [comment]
I recently attempted a version of this and hit the same problem. I think it would be best to expose the ID :)

**Comment by danielzsh** [comment]
By the way, do you have any idea why the macos tests are taking such an absurd amount of time?
image

**Comment by danielzsh** [comment]
> you’ll probably have to do a query for what is under the cursor in the active buffer to get what is being queried since the actual LSP request only contains Position information of where the cursor is at time of request

Any idea if there's an easy way to get the whole symbol based on just the cursor position?

**Comment by mikayla-maki** [comment]
I wouldn't worry about finding the name of the symbol you're hovering, leave that for a follow up PR :)

**Comment by danielzsh** [comment]
image

Here's how the message looks in its current state; it would be nice to keep a live updating count of the references that have been found, but I'm not sure how to do that (as mentioned in point 4 above).

There's also a short gap between the time the 'finding references...' status is cleared and when the references window is actually opened, since the code is currently written like so:
```rs
if status.is_some() {
  cx.update(|cx| {
      this.update(cx, |this, cx| {
          this.on_lsp_work_end(
              language_server.server_id(),
              id.to_string(),
              cx,
          );
      })
  })
  .log_err();
}
// ....
request
  .response_from_lsp(
      response,
      this.upgrade().ok_or_else(|| anyhow!("no app context"))?,
      buffer_handle,
      language_server.server_id(),
      cx,
  )
  .await
```
`cx` is moved into `response_from_lsp` so I can't move the `cx.update(...` after it. I could move the `on_lsp_work_end` logic into `response_from_lsp` directly, but this causes us to lose the generality the current approach offers (which is if we want to display the status for some LSP command, we simply need to override its `status` function). Any thoughts/suggestions?

File: crates/lsp/src/lsp.rs [hunk 1]
```diff
@@ -910,6 +910,11 @@ impl LanguageServer {
         &self.root_path
     }
 
+    /// Gets the id of the next request.
+    pub fn next_id(&self) -> &AtomicI32 {
```
Comment (mikayla-maki) [inline crates/lsp/src/lsp.rs:914]: This should just return an i32

File: crates/project/src/project.rs [hunk 2]
```diff
@@ -6612,10 +6612,30 @@ impl Project {
             let file = File::from_dyn(buffer.file()).and_then(File::as_local);
             if let (Some(file), Some(language_server)) = (file, language_server) {
                 let lsp_params = request.to_lsp(&file.abs_path(cx), buffer, &language_server, cx);
+                let status = request.status();
                 return cx.spawn(move |this, cx| async move {
                     if !request.check_capabilities(language_server.capabilities()) {
                         return Ok(Default::default());
                     }
+                    let id = language_server.next_id().load(SeqCst);
```
Comment (mikayla-maki) [inline crates/project/src/project.rs:6620]: I think this approach of grabbing the next ID seems error prone. An alternative design could be that we create a custom struct that implements `Future`, that gets returned from `language_server.request()`, that also stores the ID on itself. That way we can re-arrange this logic such that we call `language_server.request()`, get the ID off of it, and then await it after we've put in the language server progress.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]
Looks pretty good! Let's fix that ID API and see where we get to :)

**Comment by danielzsh** [comment]
I implemented a `FutureWithId` struct that contains an id in addition to wrapping a future and everything still works as before, but the implementation might need a bit more polishing? New to Rust, so I'm not too sure what's going on with all the Boxes and Pins 😓

**Comment by mikayla-maki** [comment]
Good attempt! However I had an idea for a way to do it without involving a `Box<>`, and in the process of explaining it I realized it would probably be easier to push the refactor myself :D

**Review by mikayla-maki — APPROVED** [review]

**Comment by mikayla-maki** [comment]
Thank you!

**Comment by danielzsh** [comment]
By the way, do you have any idea how we can implement either:
- Showing the actual symbol name in the status, e.g. `Finding references to AppContext` (would this perhaps require an LSP request to get the symbol name first?)
- Showing progress updates, e.g. `Found 100 references... -> Found 101 references...`

**Pull request merged**